gpt_engineer/core/chat_to_files.py

```diff
@@ -100,97 +100,104 @@ def overwrite_code_with_edits(chat: str, files_dict: FilesDict):
 @dataclass
 class Edit:
     filename: str
-    before: str
-    after: str
+    line_number: int
+    content: str
+    is_before: bool
 
 
 def parse_edits(chat: str):
     """
-    Parse edits from a chat string.
-
-    This function extracts code edits from a chat string and returns them as a list
-    of Edit objects.
-
-    Parameters
-    ----------
-    chat : str
-        The chat content containing code edits.
-
-    Returns
-    -------
-    List[Edit]
-        A list of Edit objects representing the parsed code edits.
+    Parse edits from a chat string, preserving indentation.
     """
+    edits = []
+    lines = chat.split("\n")
+    filename = ""
+    for line in lines:
+        if line.startswith("File:"):
+            filename = line.split(":")[1].strip()
+        else:
+            # Modified regex to capture leading spaces (indentation)
+            match = re.match(r"(\d+) ([-+])\s?(.*\S.*)", line)
+            if match:
+                line_number, symbol, content = match.groups()
+                line_number = int(line_number)
+                is_before = symbol == "-"
+                edits.append(Edit(filename, line_number, content, is_before))
+
+    # Sort edits such that 'before' edits come before 'after' edits
+    edits.sort(key=lambda edit: (edit.filename, edit.line_number, not edit.is_before))
+    return edits
 
-    def parse_one_edit(lines):
-        HEAD = "<<<<<<< HEAD"
-        DIVIDER = "\n=======\n"
-        UPDATE = ">>>>>>> updated"
 
-        filename = lines.pop(0)
-        text = "\n".join(lines)
-        splits = text.split(DIVIDER)
-        if len(splits) != 2:
-            raise ValueError(f"Could not parse following text as code edit: \n{text}")
-        before, after = splits
+# Helper function to compare strings ignoring spaces
+def is_similar(str1, str2):
+    # Remove spaces and convert to lowercase for case-insensitive comparison
+    str1, str2 = str1.replace(" ", "").lower(), str2.replace(" ", "").lower()
 
-        before = before.replace(HEAD, "").strip()
-        after = after.replace(UPDATE, "").strip()
+    # For single character strings, require exact match
```

Comment:
Shouldn't this part be superfluous? I believe the nice code below should imply this for all strings shorter than 10 characters? This should be added as a test case and if its passes, these three lines can be removed.